### DISCUSSION TITLE
Example

### DISCUSSION THREAD

**Discussion by iulianiuga** [hjoykim/THREE#43] [Q&A] [answered]
Hi, 
Can anybody help me with a simple example (a cube) but with a trackball camera ?
I managed migrate the first example in a form but i cannot understand from the example how enable the trackball camera and simplify the code.
Thanks  all

**Comment by hjoykim** [comment]
please show me your example source?
The  FirstSceneExample.cs  in THREEExample.Chapter1 would be the example that you request

**Reply by iulianiuga** [reply]
I used that, my question was how to make that simple new project with a simple form without using the examples tree and logic. A simple form with just one scene.

This is the code on my form....  I have just one form and one glControl


using OpenTK.Graphics.ES30;
using OpenTK.Windowing.Common;
using OpenTK.Windowing.GraphicsLibraryFramework;
using THREE;
using Color = THREE.Color;
using OpenTK.GLControl;
using System.Diagnostics;
using System.Windows.Forms;
using static System.Formats.Asn1.AsnWriter;
using Rectangle = THREE.Rectangle;
using MouseEventArgs = System.Windows.Forms.MouseEventArgs;


namespace testhree
{
    public partial class Form1 : Form
    {
    

        public GLRenderer renderer;
        public Camera camera;
        public Scene scene;
        public TrackballControls controls;
        private System.Windows.Forms.Timer _timer;
        private int timeInterval = 10;

        public Form1()
        {
            InitializeComponent();
            camera = new PerspectiveCamera();
            scene = new Scene();
        }

        private void Render()
        {
            renderer.Render(scene, camera);
            this.glControl.MakeCurrent();
            this.glControl.SwapBuffers();
        }

        void RunSample()
        {

            _timer = new System.Windows.Forms.Timer();
            _timer.Interval = timeInterval;
            _timer.Tick += (sender, e) =>
            {
                Render();
            };
            _timer.Start();
            GL.Viewport(0, 0, glControl.ClientSize.Width, glControl.ClientSize.Height);
        }

        private void Form1_Load(object sender, EventArgs e)
        {

            this.renderer = new THREE.GLRenderer();
            this.renderer.Context = glControl.Context;
            this.renderer.Width = glControl.Width;
            this.renderer.Height = glControl.Height;

            this.renderer.Init();

            InitRenderer();

            InitCamera();

            InitCameraController();

            InitLighting();


            scene.Background = Color.Hex(0x00ffff);

            var axes = new AxesHelper(20);

            scene.Add(axes);

            var planeGeometry = new PlaneGeometry(60, 20);
            var planeMaterial = new MeshBasicMaterial() { Color = Color.Hex(0xcccccc) };
            var plane = new Mesh(planeGeometry, planeMaterial);

            plane.Rotation.X = (float)(-0.5 * Math.PI);
            plane.Position.Set(15, 0, 0);

            scene.Add(plane);

            // create a cube
            var cubeGeometry = new BoxGeometry(4, 4, 4);
            var cubeMaterial = new MeshBasicMaterial() { Color = Color.Hex(0xff0000), Wireframe = true };
            var cube = new Mesh(cubeGeometry, cubeMaterial);

            // position the cube
            cube.Position.Set(-4, 3, 0);

            // add the cube to the scene

            scene.Add(cube);

            //      // create a sphere
            var sphereGeometry = new SphereGeometry(4, 20, 20);
            var sphereMaterial = new MeshBasicMaterial() { Color = Color.Hex(0x7777ff), Wireframe = true };
            var sphere = new Mesh(sphereGeometry, sphereMaterial);

            //      // position the sphere
            sphere.Position.Set(20, 4, 2);

            //      // add the sphere to the scene
            scene.Add(sphere);

        }


        public void InitCamera()
        {
            camera.Fov = 45.0f;
            camera.Aspect = this.glControl.AspectRatio;
            camera.Near = 0.1f;
            camera.Far = 1000.0f;
            camera.Position.X = -30;
            camera.Position.Y = 40;
            camera.Position.Z = 30;
            camera.LookAt(THREE.Vector3.Zero());
        }
        public void InitRenderer()
        {
            this.renderer.SetClearColor(new THREE.Color().SetHex(0xEEEEEE), 1);
            this.renderer.ShadowMap.Enabled = true;
            this.renderer.ShadowMap.type = Constants.PCFSoftShadowMap;
        }
        public void InitCameraController()
        {

            //controls = new TrackballControls(this, camera);
            //controls.StaticMoving = false;
            //controls.RotateSpeed = 4.0f;
            //controls.ZoomSpeed = 3;
            //controls.PanSpeed = 3;
            //controls.NoZoom = false;
            //controls.NoPan = false;
            //controls.NoRotate = false;
            //controls.StaticMoving = true;
            //controls.DynamicDampingFactor = 0.3f;
            //controls.Update();
        }
        public virtual void InitLighting()
        {

        }

        private void glControl_Paint(object sender, PaintEventArgs e)
        {
            Render();
        }

        void Run3d()
        {

            this.glControl.MakeCurrent();

            _timer = new System.Windows.Forms.Timer();
            _timer.Interval = timeInterval;
            _timer.Tick += (sender, e) =>
            {
                Render();
            };
            _timer.Start();
            GL.Viewport(0, 0, glControl.ClientSize.Width, glControl.ClientSize.Height);
            
        }

        private void glControl_Resize(object sender, EventArgs e)
        {
            GL.Viewport(0, 0, glControl.ClientSize.Width, glControl.ClientSize.Height);
            camera.Aspect = this.glControl.AspectRatio;
            camera.UpdateProjectionMatrix();
            if (renderer != null)
            {
                renderer.Resize(glControl.ClientSize.Width, glControl.ClientSize.Height);
                camera.Aspect = this.glControl.AspectRatio;
                camera.UpdateProjectionMatrix();
            }
        }
    }
}

**Comment by hjoykim** [comment, marked as answer]
hello
TrackballController must inherit ControlContainer.
Try it like this

```
using OpenTK.Windowing.Common;
using OpenTK.WinForms;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using THREE;
using Color = THREE.Color;
namespace SingleFormsDemo
{
    public class THREEAppInstance : ControlsContainer
    {
        public GLRenderer renderer;
        public Camera camera;
        public Scene scene;
        public TrackballControls controls;
        public GLControl glControl;

        public THREEAppInstance()
        {
            renderer = new GLRenderer();
            camera = new PerspectiveCamera();
            scene = new Scene();           
            this.renderer = new THREE.GLRenderer();
        }

        public override THREE.Rectangle GetClientRectangle()
        {
            return new THREE.Rectangle(0, 0, renderer.Width, renderer.Height);
        }
        public void InitCamera()
        {
            camera.Fov = 45.0f;
            camera.Aspect = this.glControl.AspectRatio;
            camera.Near = 0.1f;
            camera.Far = 1000.0f;
            camera.Position.X = -30;
            camera.Position.Y = 40;
            camera.Position.Z = 30;
            camera.LookAt(THREE.Vector3.Zero());
        }
        public void InitRenderer()
        {
            this.renderer.SetClearColor(new THREE.Color().SetHex(0xEEEEEE), 1);
            this.renderer.ShadowMap.Enabled = true;
            this.renderer.ShadowMap.type = Constants.PCFSoftShadowMap;
        }
        public void InitCameraController()
        {
            controls = new TrackballControls(this, camera);
            controls.StaticMoving = false;
            controls.RotateSpeed = 4.0f;
            controls.ZoomSpeed = 3;
            controls.PanSpeed = 3;
            controls.NoZoom = false;
            controls.NoPan = false;
            controls.NoRotate = false;
            controls.StaticMoving = true;
            controls.DynamicDampingFactor = 0.3f;
            controls.Update();
        }
        public void InitLighting()
        {

        }
        public void Load(GLControl control)
        {
            glControl = control;

            this.renderer.Context = control.Context;
            this.renderer.Width = control.Width;
            this.renderer.Height = control.Height;

            this.renderer.Init();

            Init();
            
        }
        public void Init()
        {
            InitRenderer();

            InitCamera();

            InitCameraController();

            InitLighting();

            scene.Background = Color.Hex(0x00ffff);

            var axes = new AxesHelper(20);

            scene.Add(axes);

            var planeGeometry = new PlaneGeometry(60, 20);
            var planeMaterial = new MeshBasicMaterial() { Color = Color.Hex(0xcccccc) };
            var plane = new Mesh(planeGeometry, planeMaterial);

            plane.Rotation.X = (float)(-0.5 * Math.PI);
            plane.Position.Set(15, 0, 0);

            scene.Add(plane);

            // create a cube
            var cubeGeometry = new BoxGeometry(4, 4, 4);
            var cubeMaterial = new MeshBasicMaterial() { Color = Color.Hex(0xff0000), Wireframe = true };
            var cube = new Mesh(cubeGeometry, cubeMaterial);

            // position the cube
            cube.Position.Set(-4, 3, 0);

            // add the cube to the scene

            scene.Add(cube);

            //      // create a sphere
            var sphereGeometry = new SphereGeometry(4, 20, 20);
            var sphereMaterial = new MeshBasicMaterial() { Color = Color.Hex(0x7777ff), Wireframe = true };
            var sphere = new Mesh(sphereGeometry, sphereMaterial);

            //      // position the sphere
            sphere.Position.Set(20, 4, 2);

            //      // add the sphere to the scene
            scene.Add(sphere);
        }
        public  void Render()
        {
            controls?.Update();
            renderer?.Render(scene, camera);
        }
        public override void OnResize(ResizeEventArgs clientSize)
        {
            if (renderer != null)
            {
                renderer.Resize(clientSize.Width, clientSize.Height);
                camera.Aspect = this.glControl.AspectRatio;
                camera.UpdateProjectionMatrix();
            }
            base.OnResize(clientSize);
        }
    }
}
```
and Form1 code are
```
using System.Windows.Forms;
using OpenTK.Graphics.ES30;
using OpenTK.Windowing.Common;
using OpenTK.Windowing.GraphicsLibraryFramework;
using OpenTK.WinForms;
using THREE;
using Keys = OpenTK.Windowing.GraphicsLibraryFramework.Keys;
namespace SingleFormsDemo
{
    public partial class Form1 : Form
    {
        public THREEAppInstance threeInstance = null;

        private System.Windows.Forms.Timer _timer;
        private int timeInterval = 10;

        public Form1()
        {
            InitializeComponent();

            glControl.MouseWheel += glControl_MouseWheel;
        }
        private void Run()
        {
            _timer = new System.Windows.Forms.Timer();
            _timer.Interval = timeInterval;
            _timer.Tick += (sender, e) =>
            {
                Render();
            };
            _timer.Start();
        }

        private void Render()
        {
            this.glControl.MakeCurrent();
            threeInstance.Render();
            this.glControl.SwapBuffers();
        }

        private void glControl_MouseWheel(object? sender, System.Windows.Forms.MouseEventArgs e)
        {
            (threeInstance.glControl as GLControl).Focus();
            threeInstance.OnMouseWheel(e.X, e.Y, e.Delta);
        }

        private void glControl_Load(object sender, EventArgs e)
        {
            this.glControl.Profile = OpenTK.Windowing.Common.ContextProfile.Compatability;
            threeInstance = new THREEAppInstance();
            threeInstance.Load(glControl);
            threeInstance.OnResize(new ResizeEventArgs(glControl.ClientSize.Width, glControl.ClientSize.Height));

            Run();
        }

        private void glControl_Resize(object sender, EventArgs e)
        {
            var control = sender as GLControl;

            if (control.ClientSize.Height == 0)
                control.ClientSize = new Size(control.ClientSize.Width, 1);

            GL.Viewport(0, 0, control.ClientSize.Width, control.ClientSize.Height);
            threeInstance?.OnResize(new ResizeEventArgs(control.ClientSize.Width, control.ClientSize.Height));
        }

        private void glControl_Paint(object sender, PaintEventArgs e)
        {
            Render();
        }
        private MouseButton GetMouseButton(System.Windows.Forms.MouseEventArgs e)
        {
            MouseButton button = MouseButton.Left;
            switch (e.Button)
            {
                case MouseButtons.Middle:
                    button = MouseButton.Middle;
                    break;
                case MouseButtons.Right:
                    button = MouseButton.Right;
                    break;
                case MouseButtons.Left:
                case MouseButtons.None:
                default:
                    break;
            }
            return button;
        }

        private void glControl_MouseDown(object sender, System.Windows.Forms.MouseEventArgs e)
        {
            threeInstance?.OnMouseDown(GetMouseButton(e), e.X, e.Y);
        }

        private void glControl_MouseMove(object sender, System.Windows.Forms.MouseEventArgs e)
        {
            threeInstance?.OnMouseMove(GetMouseButton(e), e.X, e.Y);
        }

        private void glControl_MouseUp(object sender, System.Windows.Forms.MouseEventArgs e)
        {
            threeInstance?.OnMouseUp(GetMouseButton(e), e.X, e.Y);
        }

        private void glControl_KeyUp(object sender, KeyEventArgs e)
        {
            Keys key = (Keys)e.KeyCode;
            switch (e.KeyCode)
            {
                case System.Windows.Forms.Keys.Right:
                    key = Keys.Right;
                    break;
                case System.Windows.Forms.Keys.Left:
                    key = Keys.Left;
                    break;
                case System.Windows.Forms.Keys.Down:
                    key = Keys.Down;
                    break;
                case System.Windows.Forms.Keys.Up:
                    key = Keys.Up;
                    break;

            }
            threeInstance?.OnKeyUp(key, e.KeyValue, (KeyModifiers)e.Modifiers);
        }

        private void glControl_KeyDown(object sender, KeyEventArgs e)
        {
            Keys key = (Keys)e.KeyCode;
            switch (e.KeyCode)
            {
                case System.Windows.Forms.Keys.Right:
                    key = Keys.Right;
                    break;
                case System.Windows.Forms.Keys.Left:
                    key = Keys.Left;
                    break;
                case System.Windows.Forms.Keys.Down:
                    key = Keys.Down;
                    break;
                case System.Windows.Forms.Keys.Up:
                    key = Keys.Up;
                    break;

            }
            threeInstance?.OnKeyDown(key, e.KeyValue, (KeyModifiers)e.Modifiers);
        }

        private void glControl_KeyPress(object sender, System.Windows.Forms.KeyPressEventArgs e)
        {
            threeInstance?.OnKeyPress(e.KeyChar.ToString());
        }
    }
}
```

I attacted this project to zip
[SingleFormsDemo.zip](https://github.com/user-attachments/files/17335846/SingleFormsDemo.zip)

**Reply by iulianiuga** [reply]
Thank you for your help, works like a charm!
